### PR TITLE
Lazy loading

### PR DESCRIPTION
**Pull request by YuliiaOsadchuk**

### CODE REVIEW

File: src/Router.tsx [hunk 1]
```diff
@@ -1,16 +1,19 @@
-import React from 'react';
+import React, { lazy, Suspense } from 'react';
 import { BrowserRouter, Route, Routes } from 'react-router-dom';
 
-import Signin from './components/signin/signin';
 import { APP_ROUTES } from './enums';
-import Home from './pages/home/home';
+
+const Signin = lazy(() => import('./components/signin/signin'));
```
Comment (b-sht-lionwood) [inline src/Router.tsx]: fix pipeline errors
Screenshot 2021-12-07 at 18 48 48

File: src/Router.tsx [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+const Home = lazy(() => import('./pages/home/home'));
 
 const Router: React.FC = () => (
   <BrowserRouter>
-    <Routes>
-      <Route path={APP_ROUTES.SIGNIN} element={<Signin />} />
-      <Route path={APP_ROUTES.HOME} element={<Home />} />
-    </Routes>
+    <Suspense fallback={<div>Download...</div>}>
```
Comment (b-sht-lionwood) [inline src/Router.tsx]: use material-ui loader here

**Review by b-sht-lionwood — CHANGES_REQUESTED** [review]

File: src/Router.tsx (hunk 1 above)
Comment (b-sht-lionwood) [inline src/Router.tsx]: create new file `sign-in-lazy.tsx` in same folder where component, where you will export lazy loaded component

File: src/Router.tsx [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
+      <Routes>
+        <Route path={APP_ROUTES.SIGNIN} element={<Signin />} />
```
Comment (b-sht-lionwood) [inline src/Router.tsx]: is is better to suspense each route separately.

You can create such high-ordero-component for that:
```
interface DtWithSuspenseProps {
  Component?: LazyExoticComponent<ComponentType>;
  loader?: ReactElement;
}

const DEFAULT_LOADER = <DtSkeleton variant='rectangular' />;

/**
 * This HOC intended to wrap components that you want to lazy load {@link https://reactjs.org/docs/code-splitting.html}
 *
 * @param {React.LazyExoticComponent<React.FC>} Component Component that you want to exclude from bundle and lazy load on demand
 * @param {React.ReactElement} loader Component that will be displayed till component load
 *
 * @example const componentToBeLazyLoaded = withSuspense(lazy(() => import(pathToComponent)));
 */
export const dtWithSuspense =
  (Component: LazyExoticComponent<FC>, loader: ReactElement | undefined = DEFAULT_LOADER): FC<DtWithSuspenseProps> =>
  (props): ReactElement => {
    return (
      <Suspense fallback={loader}>
        <Component {...props} />
      </Suspense>
    );
  };
```

**Review by b-sht-lionwood — CHANGES_REQUESTED** [review]

**Comment by YuliiaOsadchuk** [comment]
Download with slow 3g internet

https://user-images.githubusercontent.com/70544391/145364820-4e34ee2c-cdb9-4654-8a0e-e546a80b8ba1.mov

File: src/Router.tsx [hunk 4]
```diff
@@ -1,15 +1,15 @@
 import React from 'react';
```
Comment (b-sht-lionwood) [inline src/Router.tsx:1]: file name should be lowercase

File: src/route-with-suspense.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+export const routeWithSuspense =
+  (Component: LazyExoticComponent<FC>, Loader: FC): FC<RouteWithSuspenseProps> =>
+  (props): ReactElement =>
+    (
+      <Suspense fallback={<Loader />}>
+        <Component {...props} />
+      </Suspense>
+    );
```
Comment (b-sht-lionwood) [inline src/route-with-suspense.tsx]: add jsdoc documnetation here of how to use this component

File: src/route-with-suspense.tsx [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import React, { Suspense, FC, LazyExoticComponent, ReactElement, ComponentType } from 'react';
+
+interface RouteWithSuspenseProps {
+  Component?: LazyExoticComponent<ComponentType>;
+  loader?: ReactElement;
+}
+
+export const routeWithSuspense =
```
Comment (b-sht-lionwood) [inline src/route-with-suspense.tsx]:
```suggestion
export const withSuspense =
```

File: src/pages/home-page/home-page.tsx [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+
+import Home from '../../components/home/home';
+
+const HomePage: React.FC = () => <Home />;
```
Comment (b-sht-lionwood) [inline src/pages/home-page/home-page.tsx]:
```suggestion
const HomePage = () => <Home />;
```

File: src/components/signin/sign-in.tsx [hunk 8]
```diff
@@ -14,7 +14,7 @@ import Copyright from '../copyrigth/copyrigth';
 import { FlexColumn, FlexCenter, SignInButton } from './sign-in.styles';
 import validationSchema from './validationSchema';
 
-const Signin: React.FC = () => {
+const Signin: FC = () => {
```
Comment (b-sht-lionwood) [inline src/components/signin/sign-in.tsx]:
```suggestion
const Signin = () => {
```

File: src/components/home/home-lazy.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import { FC, lazy, LazyExoticComponent } from 'react';
+
+import { routeWithSuspense } from '../../route-with-suspense';
+import Loader from '../loader/loader';
+
+const HomeLazy: LazyExoticComponent<FC> = lazy(() => import('./home'));
```
Comment (b-sht-lionwood) [inline src/components/home/home-lazy.ts]: No need to specify such file type
```suggestion
const HomeLazy = lazy(() => import('./home'));
```

File: src/route-with-suspense.tsx [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import React, { Suspense, FC, LazyExoticComponent, ReactElement, ComponentType } from 'react';
+
+interface RouteWithSuspenseProps {
+  Component?: LazyExoticComponent<ComponentType>;
+  loader?: ReactElement;
+}
+
```
Comment (b-sht-lionwood) [inline src/route-with-suspense.tsx]: add default loader

File: src/components/signin/sign-in-lazy.tsx [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import { FC, lazy, LazyExoticComponent } from 'react';
+
+import { routeWithSuspense } from '../../route-with-suspense';
+import Loader from '../loader/loader';
+
+const SigninLazy: LazyExoticComponent<FC> = lazy(() => import('./sign-in'));
```
Comment (b-sht-lionwood) [inline src/components/signin/sign-in-lazy.tsx]: No need to specify such file type
```suggestion
const SigninLazy = lazy(() => import('./sign-in'));
```

**Review by b-sht-lionwood — CHANGES_REQUESTED** [review]

File: src/components/home/home-lazy.tsx [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import React, { lazy } from 'react';
+
+import { withSuspense } from '../../with-suspense';
+import Progress from '../progress/progress';
+
+const HomeLazy = lazy(() => import('./home'));
+
+export default withSuspense(HomeLazy, <Progress />);
```
Comment (b-sht-lionwood) [inline src/components/home/home-lazy.tsx]: you already have same default loader inside function

File: src/components/signin/sign-in-lazy.tsx [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import React, { lazy } from 'react';
+
+import { withSuspense } from '../../with-suspense';
+import Progress from '../progress/progress';
+
+const SigninLazy = lazy(() => import('./sign-in'));
+
+export default withSuspense(SigninLazy, <Progress />);
```
Comment (b-sht-lionwood) [inline src/components/signin/sign-in-lazy.tsx]: you already have same default loader inside function

**Review by b-sht-lionwood — CHANGES_REQUESTED** [review]

**Review by b-sht-lionwood — APPROVED** [review]

**Pull request merged**